### PR TITLE
Improve Contour2D.cut_by_wire

### PR DESCRIPTION
**Pull request by sbendjebla**
* **Please check if the PR fulfills these requirements**
- [ ] The commit messages follows our guidelines (explicit, in english)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] CHANGELOG.md has been updated

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [ ] Bug fix:
- [ ] New features:
- [ ] Documentation
- [x] Optimization

* **Other information**:

I improved `cut_by_wire`, based on the new version of `cut_by_line`. (_these two methods could be refactorized after_)

I developped some related methods: `Wire.sort_points_along_wire`, `Wire2D.linesegment_crossings`, `Wire2D.wire_crossings`
and modified: `Wire2D.wire_intersections`, `LineSegment.split`

I added a script to illustrate 
image

### CODE REVIEW

File: volmdlr/wires.py [hunk 1]
```diff
@@ -437,174 +441,20 @@ def wire_intersections(self, wire):
...
-        # ax.set_aspect('auto')
-        raise NotImplementedError(
-            '{} intersections not supported yet'.format(len(intersections)))
+                if intersections_points:
+                    if a[0][0] not in intersections_points:
+                        intersections.append([a[0][0], a[0][1]])
+                        intersections_points.append(a[0][0])
+                else:
+                    intersections.append([a[0][0], a[0][1]])
+                    intersections_points.append(a[0][0])
```
Comment (AxelRinghausen) [inline volmdlr/wires.py:455]:
```suggestion
                if a[0][0] not in intersections_points:
                        intersections.append([a[0][0], a[0][1]])
                        intersections_points.append(a[0][0])
```

File: volmdlr/wires.py [hunk 2]
```diff
@@ -617,8 +467,49 @@ def from_points(cls, points: List[volmdlr.Point2D]):
...
+            if a:
+                if crossings_points:
+                    if a[0][0] not in crossings_points:
+                        crossings.append([a[0][0], a[0][1]])
+                        crossings_points.append(a[0][0])
+                else:
+                    crossings.append([a[0][0], a[0][1]])
+                    crossings_points.append(a[0][0])
+
+        return crossings
```
Comment (AxelRinghausen) [inline volmdlr/wires.py:512]: I'd suggest something like that. I don't understand the first return inside the _for_ loop. 
This has to be tested, I'm not a 100% sure about what I wrote. 
```suggestion
    def wire_crossings(self, wire):
        '''
        compute crossings between two wire2d

        Parameters
        ----------
        wire : volmdlr.wires.Wire2D

        Returns
        -------
        crossings : List[(volmdlr.Point2D, volmdlr.Primitive2D)]
        '''
        crossings, crossings_points = [], []
        for primitive in wire.primitives:
            a = self.linesegment_crossings(primitive)
            if a:
                if a[0][0] not in crossings_points:
                    crossings.append([a[0][0], a[0][1]])
                    crossings_points.append(a[0][0])
        return crossings
```

**Review by AxelRinghausen — CHANGES_REQUESTED** [review]

File: scripts/contours/contour_cut_by_wire.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import volmdlr as vm
+import volmdlr.wires
+import matplotlib.pyplot as plt
+
+# %% Contour2d
+
+p = [vm.Point2D(-0.3, -0.2), vm.Point2D(0.3, -0.2),
+      vm.Point2D(0.2, 0.2), vm.Point2D(0, 0.3), vm.Point2D(-0.2, 0.2)]
+
+contour = vm.wires.Contour2D(vm.wires.ClosedPolygon2D(p).primitives)
```
[the remark this answers is not in the document]
Comment (sbendjebla) [reply scripts/contours/contour_cut_by_wire.py:20]: Done!

File: volmdlr/wires.py (hunk 1 above)
Comment (sbendjebla) [reply volmdlr/wires.py:455]: Done!

File: volmdlr/wires.py (hunk 2 above)
Comment (sbendjebla) [reply volmdlr/wires.py:512]: The first return inside the for loop, is when the primitive that define the wire crosses the contour twice in different places

image

If we use your version, we'll have just one crossing point !

File: volmdlr/wires.py [hunk 4]
```diff
@@ -1793,7 +1684,64 @@ def merge_with(self, contour2d):
...
+
+            point1 = sorted_points[cutting_points_counter]
+            point2 = sorted_points[cutting_points_counter + 1]
+
+            closing_wire = wire.extract_without_primitives(point1, point2, True)
+            closing_wire_prim = []
+            for i in range(0, len(closing_wire)):
+                if closing_wire[i]==[]:
+                    continue
+                closing_wire_prim.append(closing_wire[i])
```
[the remark this answers is not in the document]
Comment (sbendjebla) [reply volmdlr/wires.py:1723]: Thanks!

**Review by AxelRinghausen — DISMISSED** [review]

**Review by AxelRinghausen — APPROVED** [review]

**Pull request merged**